core/src/main/scala/org/apache/spark/sql/rapids/tool/profiling/ApplicationInfo.scala

```diff
@@ -181,6 +184,122 @@ object SparkPlanInfoWithStage {
   }
 }
 
+object GpuSparkPlanGraph {
+
+  /**
+   * This code is mostly copied from org.apache.spark.sql.execution.ui.SparkPlanGraph
+   * with additions/changes to fit our needs.
+   *
+   * Build a SparkPlanGraph from the root of a SparkPlan tree.
+   */
+  def apply(planInfo: SparkPlanInfo): SparkPlanGraph = {
+    val nodeIdGenerator = new AtomicLong(0)
+    val nodes = mutable.ArrayBuffer[SparkPlanGraphNode]()
+    val edges = mutable.ArrayBuffer[SparkPlanGraphEdge]()
+    val exchanges = mutable.HashMap[SparkPlanInfo, SparkPlanGraphNode]()
+    buildSparkPlanGraphNode(planInfo, nodeIdGenerator, nodes, edges, null, null, exchanges)
+    new SparkPlanGraph(nodes.toSeq, edges.toSeq)
+  }
+
+  private def buildSparkPlanGraphNode(planInfo: SparkPlanInfo,
+      nodeIdGenerator: AtomicLong,
+      nodes: mutable.ArrayBuffer[SparkPlanGraphNode],
+      edges: mutable.ArrayBuffer[SparkPlanGraphEdge],
+      parent: SparkPlanGraphNode,
+      subgraph: SparkPlanGraphCluster,
+      exchanges: mutable.HashMap[SparkPlanInfo, SparkPlanGraphNode]
+  ): Unit = {
+
+    planInfo.nodeName match {
+      case name if name.startsWith("WholeStageCodegen") =>
```

Comment:
there is code very similar to this in /profiling/GenerateDot.scala.

Can the two be commonized?  ie it might have a similar issue

Reply:
That class seems to be doing more stuff. Maybe @GaryShen2008 would have some idea if we can unify them?

Reply:
Maybe it was @nartal1?

Reply:
@andygrove Niranjan said you worked on this. Any ideas on unifying the two?

Reply:
This code is mainly identical to GenerateDot (clearly it was based on this). One difference is that this version takes stages into account. It would make sense to me to add this functionality to GenerateDot (or factor out the common code, at least).

Reply:
Are you suggesting we add the additional `case` statements for `GpuSubquery`s to `GenerateDot`?

Reply:
It seems like we shouldn't duplicate all of this code, ideally. I will take a closer look in the morning and try and make some concrete suggestions.

Reply:
I took another look. The code for `GpuSparkPlanGraph` and `SparkPlanGraph` is very similar, but the latter also contains code specific to Dot diagrams. It may be worth refactoring the GenerateDot code to separate the graph structure and the Dot generation code and then the same graph could be used for both purposes, but I'll leave it to those closer to this code to determine if it is worth the effort.